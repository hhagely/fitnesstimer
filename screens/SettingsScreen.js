import React, { Component } from 'react';
import {
	Picker,
	Text,
	View,
	ScrollView,
	Platform,
	StyleSheet,
	Switch,
	Button,
	AsyncStorage
} from 'react-native';

class SettingsScreen extends Component {
	static navigationOptions = {
		title: 'Timer Settings'
	};

	constructor(props) {
		super(props);

		this.state = {
			timerType: 'AMRAP',
			timerDuration: 10,
			countdown: true,
			countdownDuration: 10
		};
	}

	renderEmomSelector() {
		if (this.state.timerType === 'EMOM') {
			return (
				<View>
					<Text>EMOM Style</Text>

					<Picker
						selectedValue={this.state.emomStyle}
						onValueChange={(itemValue, itemIndex) =>
							this.setState({ emomStyle: itemValue })
						}
					>
						<Picker.Item label="Every minute" value={1} />
						<Picker.Item label="Every 2 minutes" value={2} />
						<Picker.Item label="Every 3 minutes" value={3} />
					</Picker>
				</View>
			);
		}
	}

	generatePickerItems() {
		return Array(99)
			.fill()
			.map((entry, index) => {
				return (
					<Picker.Item
						key={index + 1}
						label={index === 0 ? `1 minute` : `${index + 1} minutes`}
						value={index + 1}
					/>
				);
			});
	}

	renderCountdownDurationPicker() {
		if (this.state.countdown) {
			return (
				<View>
					<Text>Countdown Duration</Text>
					<Picker
						selectedValue={this.state.countdownDuration}
						onValueChange={(itemValue, itemIndex) => {
							this.setState({ countdownDuration: itemValue });
						}}
					>
						<Picker.Item label="1 second" value={1} />
						<Picker.Item label="2 seconds" value={2} />
						<Picker.Item label="3 seconds" value={3} />
						<Picker.Item label="4 seconds" value={4} />
						<Picker.Item label="5 seconds" value={5} />
						<Picker.Item label="6 seconds" value={6} />
						<Picker.Item label="7 seconds" value={7} />
						<Picker.Item label="8 seconds" value={8} />
						<Picker.Item label="9 seconds" value={9} />
						<Picker.Item label="10 seconds" value={10} />
					</Picker>
				</View>
			);
		}
	}

	setTimerAndNavigate() {
		// save data in async storage and then navigate to the home screen.
		let timerSettings = {
			timerType: this.state.timerType,
			timerDuration: this.state.timerDuration,
			countdown: this.state.countdown,
			countdownDuration: this.state.countdownDuration
		};

		if (timerSettings.timerType === 'EMOM') {
			timerSettings.emomStyle = this.state.emomStyle;
		}

		this.props.navigation.navigate('Home', {
			timerSettings
		});
	}

	render() {
		return (
			<View style={styles.container}>
				<ScrollView
					style={styles.container}
					contentContainerStyle={styles.contentContainer}
				>
					<View>
						<Text>Timer Type</Text>
						<Picker
							selectedValue={this.state.timerType}
							onValueChange={(itemValue, itemIndex) => {
								this.setState({ timerType: itemValue });
								if (itemValue === 'Tabata' || itemValue === 'ReverseTabata') {
									this.setState({ timerDuration: 4 });
								}
							}}
						>
							<Picker.Item label="EMOM" value="EMOM" />
							<Picker.Item label="AMRAP" value="AMRAP" />
							<Picker.Item label="Tabata" value="Tabata" />
							<Picker.Item label="Reverse Tabata" value="ReverseTabata" />
						</Picker>
					</View>
					{this.renderEmomSelector()}
					<View>
						<Text>Timer Duration</Text>
						<Picker
							selectedValue={this.state.timerDuration}
							onValueChange={(itemValue, itemIndex) => {
								this.setState({ timerDuration: itemValue });
							}}
						>
							{this.generatePickerItems()}
						</Picker>
					</View>
					<View>
						<Text>Countdown</Text>
						<Switch
							value={this.state.countdown}
							onValueChange={(value) => {
								this.setState({ countdown: value });
							}}
						/>
					</View>
					{this.renderCountdownDurationPicker()}
					<View style={{ marginTop: 16 }}>
						<Button
							title="Set Timer"
							onPress={() => this.setTimerAndNavigate()}
						/>
					</View>
				</ScrollView>
			</View>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: '#fff'
	},
	contentContainer: {
		paddingTop: 30,
		paddingHorizontal: 12
	}
});

export default SettingsScreen;
