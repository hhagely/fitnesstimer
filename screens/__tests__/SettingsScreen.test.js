import { Button, Picker, Switch } from 'react-native';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
// import configureStore from 'redux-mock-store';

import { SettingsScreen } from '../SettingsScreen';
// import reducers from '../../reducers';

// const middlewares = [];
// const mockStore = configureStore(middlewares);

configure({ adapter: new Adapter() });

const initialState = {
	timerType: 'AMRAP',
	timerDuration: 10,
	countdown: true,
	countdownDuration: 10
};

const tempProps = {
	saveTimerSettings: jest.fn(),
	navigation: {
		navigate: jest.fn()
	}
};

let wrapper;

beforeEach(() => {
	wrapper = shallow(<SettingsScreen {...tempProps} />);

	// wrapper = shallow(<SettingsScreen {...tempProps} />, {
	// 	context: { store: mockStore(initialState) }
	// });
});

describe('<SettingsScreen />', () => {
	it('Renders the SettingsScreen component', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('calls componentDidMount', () => {
		jest.spyOn(SettingsScreen.prototype, 'componentDidMount');

		wrapper = shallow(<SettingsScreen {...tempProps} />);
		expect(SettingsScreen.prototype.componentDidMount.mock.calls.length).toBe(
			1
		);
	});

	it('Successfully changes the timerType', () => {
		const picker = wrapper.find(Picker).at(0);

		picker.simulate('valueChange', { itemValue: 'AMRAP' });

		expect(picker.props().selectedValue).toBe('AMRAP');
	});

	it('renders 4 pickers when timer type is EMOM', () => {
		const toggle = wrapper.find(Switch).at(0);

		toggle.simulate('valueChange', true);

		wrapper.update();

		const picker = wrapper.find(Picker).at(0);

		picker.simulate('valueChange', 'EMOM');

		wrapper.update();

		expect(wrapper.find(Picker).length).toBe(4);
	});

	it('calls Switch onvalueChange', () => {
		const toggle = wrapper.find(Switch).at(0);

		toggle.simulate('valueChange', true);

		expect(wrapper.state('countdown')).toBe(true);
	});

	it('Successfully changes the countdownDuration', () => {
		const toggle = wrapper.find(Switch).at(0);

		toggle.simulate('valueChange', true);

		wrapper.update();

		const picker = wrapper.find(Picker).at(2);

		picker.simulate('valueChange', 10);

		expect(wrapper.state('timerDuration')).toBe(10);
	});

	it('successfully changes the emom style picker', () => {
		const timerType = wrapper.find(Picker).at(0);

		timerType.simulate('valueChange', 'EMOM');

		wrapper.update();

		const emomStyle = wrapper.find(Picker).at(1);

		emomStyle.simulate('valueChange', 2);

		expect(wrapper.state('emomStyle')).toBe(2);
	});

	it('contains a button to be clicked', () => {
		const button = wrapper.find(Button).at(0);

		expect(button.exists()).toEqual(true);
	});

	it('calls setTimerAndNavigate when button is clicked', () => {
		const button = wrapper.find(Button).at(0);
		button.simulate('press');

		expect(tempProps.saveTimerSettings).toHaveBeenCalled();
	});
});
