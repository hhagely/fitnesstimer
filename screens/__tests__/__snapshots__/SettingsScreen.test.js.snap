// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsScreen /> Renders the SettingsScreen component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <ScrollView
    contentContainerStyle={
      Object {
        "paddingHorizontal": 12,
        "paddingTop": 30,
      }
    }
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Timer Type
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue="AMRAP"
      >
        <PickerItem
          label="EMOM"
          value="EMOM"
        />
        <PickerItem
          label="AMRAP"
          value="AMRAP"
        />
        <PickerItem
          label="Tabata"
          value="Tabata"
        />
        <PickerItem
          label="Reverse Tabata"
          value="ReverseTabata"
        />
      </Picker>
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Timer Duration
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue={10}
      >
        <PickerItem
          key="1"
          label="1 minute"
          value={1}
        />
        <PickerItem
          key="2"
          label="2 minutes"
          value={2}
        />
        <PickerItem
          key="3"
          label="3 minutes"
          value={3}
        />
        <PickerItem
          key="4"
          label="4 minutes"
          value={4}
        />
        <PickerItem
          key="5"
          label="5 minutes"
          value={5}
        />
        <PickerItem
          key="6"
          label="6 minutes"
          value={6}
        />
        <PickerItem
          key="7"
          label="7 minutes"
          value={7}
        />
        <PickerItem
          key="8"
          label="8 minutes"
          value={8}
        />
        <PickerItem
          key="9"
          label="9 minutes"
          value={9}
        />
        <PickerItem
          key="10"
          label="10 minutes"
          value={10}
        />
        <PickerItem
          key="11"
          label="11 minutes"
          value={11}
        />
        <PickerItem
          key="12"
          label="12 minutes"
          value={12}
        />
        <PickerItem
          key="13"
          label="13 minutes"
          value={13}
        />
        <PickerItem
          key="14"
          label="14 minutes"
          value={14}
        />
        <PickerItem
          key="15"
          label="15 minutes"
          value={15}
        />
        <PickerItem
          key="16"
          label="16 minutes"
          value={16}
        />
        <PickerItem
          key="17"
          label="17 minutes"
          value={17}
        />
        <PickerItem
          key="18"
          label="18 minutes"
          value={18}
        />
        <PickerItem
          key="19"
          label="19 minutes"
          value={19}
        />
        <PickerItem
          key="20"
          label="20 minutes"
          value={20}
        />
        <PickerItem
          key="21"
          label="21 minutes"
          value={21}
        />
        <PickerItem
          key="22"
          label="22 minutes"
          value={22}
        />
        <PickerItem
          key="23"
          label="23 minutes"
          value={23}
        />
        <PickerItem
          key="24"
          label="24 minutes"
          value={24}
        />
        <PickerItem
          key="25"
          label="25 minutes"
          value={25}
        />
        <PickerItem
          key="26"
          label="26 minutes"
          value={26}
        />
        <PickerItem
          key="27"
          label="27 minutes"
          value={27}
        />
        <PickerItem
          key="28"
          label="28 minutes"
          value={28}
        />
        <PickerItem
          key="29"
          label="29 minutes"
          value={29}
        />
        <PickerItem
          key="30"
          label="30 minutes"
          value={30}
        />
        <PickerItem
          key="31"
          label="31 minutes"
          value={31}
        />
        <PickerItem
          key="32"
          label="32 minutes"
          value={32}
        />
        <PickerItem
          key="33"
          label="33 minutes"
          value={33}
        />
        <PickerItem
          key="34"
          label="34 minutes"
          value={34}
        />
        <PickerItem
          key="35"
          label="35 minutes"
          value={35}
        />
        <PickerItem
          key="36"
          label="36 minutes"
          value={36}
        />
        <PickerItem
          key="37"
          label="37 minutes"
          value={37}
        />
        <PickerItem
          key="38"
          label="38 minutes"
          value={38}
        />
        <PickerItem
          key="39"
          label="39 minutes"
          value={39}
        />
        <PickerItem
          key="40"
          label="40 minutes"
          value={40}
        />
        <PickerItem
          key="41"
          label="41 minutes"
          value={41}
        />
        <PickerItem
          key="42"
          label="42 minutes"
          value={42}
        />
        <PickerItem
          key="43"
          label="43 minutes"
          value={43}
        />
        <PickerItem
          key="44"
          label="44 minutes"
          value={44}
        />
        <PickerItem
          key="45"
          label="45 minutes"
          value={45}
        />
        <PickerItem
          key="46"
          label="46 minutes"
          value={46}
        />
        <PickerItem
          key="47"
          label="47 minutes"
          value={47}
        />
        <PickerItem
          key="48"
          label="48 minutes"
          value={48}
        />
        <PickerItem
          key="49"
          label="49 minutes"
          value={49}
        />
        <PickerItem
          key="50"
          label="50 minutes"
          value={50}
        />
        <PickerItem
          key="51"
          label="51 minutes"
          value={51}
        />
        <PickerItem
          key="52"
          label="52 minutes"
          value={52}
        />
        <PickerItem
          key="53"
          label="53 minutes"
          value={53}
        />
        <PickerItem
          key="54"
          label="54 minutes"
          value={54}
        />
        <PickerItem
          key="55"
          label="55 minutes"
          value={55}
        />
        <PickerItem
          key="56"
          label="56 minutes"
          value={56}
        />
        <PickerItem
          key="57"
          label="57 minutes"
          value={57}
        />
        <PickerItem
          key="58"
          label="58 minutes"
          value={58}
        />
        <PickerItem
          key="59"
          label="59 minutes"
          value={59}
        />
        <PickerItem
          key="60"
          label="60 minutes"
          value={60}
        />
        <PickerItem
          key="61"
          label="61 minutes"
          value={61}
        />
        <PickerItem
          key="62"
          label="62 minutes"
          value={62}
        />
        <PickerItem
          key="63"
          label="63 minutes"
          value={63}
        />
        <PickerItem
          key="64"
          label="64 minutes"
          value={64}
        />
        <PickerItem
          key="65"
          label="65 minutes"
          value={65}
        />
        <PickerItem
          key="66"
          label="66 minutes"
          value={66}
        />
        <PickerItem
          key="67"
          label="67 minutes"
          value={67}
        />
        <PickerItem
          key="68"
          label="68 minutes"
          value={68}
        />
        <PickerItem
          key="69"
          label="69 minutes"
          value={69}
        />
        <PickerItem
          key="70"
          label="70 minutes"
          value={70}
        />
        <PickerItem
          key="71"
          label="71 minutes"
          value={71}
        />
        <PickerItem
          key="72"
          label="72 minutes"
          value={72}
        />
        <PickerItem
          key="73"
          label="73 minutes"
          value={73}
        />
        <PickerItem
          key="74"
          label="74 minutes"
          value={74}
        />
        <PickerItem
          key="75"
          label="75 minutes"
          value={75}
        />
        <PickerItem
          key="76"
          label="76 minutes"
          value={76}
        />
        <PickerItem
          key="77"
          label="77 minutes"
          value={77}
        />
        <PickerItem
          key="78"
          label="78 minutes"
          value={78}
        />
        <PickerItem
          key="79"
          label="79 minutes"
          value={79}
        />
        <PickerItem
          key="80"
          label="80 minutes"
          value={80}
        />
        <PickerItem
          key="81"
          label="81 minutes"
          value={81}
        />
        <PickerItem
          key="82"
          label="82 minutes"
          value={82}
        />
        <PickerItem
          key="83"
          label="83 minutes"
          value={83}
        />
        <PickerItem
          key="84"
          label="84 minutes"
          value={84}
        />
        <PickerItem
          key="85"
          label="85 minutes"
          value={85}
        />
        <PickerItem
          key="86"
          label="86 minutes"
          value={86}
        />
        <PickerItem
          key="87"
          label="87 minutes"
          value={87}
        />
        <PickerItem
          key="88"
          label="88 minutes"
          value={88}
        />
        <PickerItem
          key="89"
          label="89 minutes"
          value={89}
        />
        <PickerItem
          key="90"
          label="90 minutes"
          value={90}
        />
        <PickerItem
          key="91"
          label="91 minutes"
          value={91}
        />
        <PickerItem
          key="92"
          label="92 minutes"
          value={92}
        />
        <PickerItem
          key="93"
          label="93 minutes"
          value={93}
        />
        <PickerItem
          key="94"
          label="94 minutes"
          value={94}
        />
        <PickerItem
          key="95"
          label="95 minutes"
          value={95}
        />
        <PickerItem
          key="96"
          label="96 minutes"
          value={96}
        />
        <PickerItem
          key="97"
          label="97 minutes"
          value={97}
        />
        <PickerItem
          key="98"
          label="98 minutes"
          value={98}
        />
        <PickerItem
          key="99"
          label="99 minutes"
          value={99}
        />
      </Picker>
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Countdown
      </Text>
      <Switch
        disabled={false}
        onValueChange={[Function]}
        value={true}
      />
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Countdown Duration
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue={10}
      >
        <PickerItem
          label="1 second"
          value={1}
        />
        <PickerItem
          label="2 seconds"
          value={2}
        />
        <PickerItem
          label="3 seconds"
          value={3}
        />
        <PickerItem
          label="4 seconds"
          value={4}
        />
        <PickerItem
          label="5 seconds"
          value={5}
        />
        <PickerItem
          label="6 seconds"
          value={6}
        />
        <PickerItem
          label="7 seconds"
          value={7}
        />
        <PickerItem
          label="8 seconds"
          value={8}
        />
        <PickerItem
          label="9 seconds"
          value={9}
        />
        <PickerItem
          label="10 seconds"
          value={10}
        />
      </Picker>
    </View>
    <View
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <Button
        onPress={[Function]}
        title="Set Timer"
      />
    </View>
  </ScrollView>
</View>
`;
